xc/common/utils/prjxray_form_channels.py

```diff
@@ -2159,6 +2107,164 @@ def create_vpr_grid(conn, grid_map_output):
+        dest_wire = list(node_to_site_pins(conn, dest_node_pkey))
+        assert len(dest_wire) == 1
+        _, dest_tile_pkey, dest_wire_in_tile_pkey = dest_wire[0]
+
+        cur2.execute(
+            """
+SELECT tile_type_pkey, grid_x, grid_y FROM tile WHERE pkey = ?;""",
+            (dest_tile_pkey, )
+        )
+
+        res = cur2.fetchone()
+        dest_tile_type_pkey, dest_grid_x, dest_grid_y = res
+
+        if dest_grid_x == src_grid_x and dest_grid_y == src_grid_y:
+            tile = new_grid[(dest_grid_x, dest_grid_y)]
+
+            cur2.execute(
+                "SELECT site_pkey FROM wire_in_tile WHERE pkey = ?",
+                (src_wire_in_tile_pkey, )
+            )
+            src_site_pkey = cur2.fetchone()[0]
+
+            cur2.execute(
+                "SELECT site_pkey FROM wire_in_tile WHERE pkey = ?",
+                (dest_wire_in_tile_pkey, )
+            )
+            dest_site_pkey = cur2.fetchone()[0]
+
+            src_site_instance = None
+            dest_site_instance = None
+            for site in tile.sites:
+                cur2.execute(
+                    "SELECT name FROM site_instance WHERE site_pkey = ? AND phy_tile_pkey = ?"
+                    "", (
+                        site.site_pkey,
+                        site.phy_tile_pkey,
+                    )
+                )
+
+                res = cur2.fetchone()[0]
+
+                if site.site_pkey == src_site_pkey:
+                    src_site_instance = res
+                elif site.site_pkey == dest_site_pkey:
+                    dest_site_instance = res
+
+            assert src_site_instance and dest_site_instance
+
+            sites_with_direct[src_site_instance] = dest_site_instance
+
+    # Build the Canonical to VPR grid mapping
+    for (grid_x, grid_y), tile in new_grid.items():
+        for site in tile.sites:
+            cur.execute(
+                "SELECT name FROM site_instance WHERE site_pkey = ? AND phy_tile_pkey = ?",
+                (
+                    site.site_pkey,
+                    site.phy_tile_pkey,
+                )
+            )
+
+            site_name = cur.fetchone()[0]
+
+            cur.execute(
+                "SELECT name, clock_region_pkey, grid_x, grid_y FROM phy_tile WHERE pkey = ?",
+                (site.phy_tile_pkey, )
+            )
+
+            phy_tile_name, clock_region, canon_x, canon_y = cur.fetchone()
+
+            cur.execute(
+                "SELECT name FROM site_type WHERE pkey = ?",
+                (site.site_type_pkey, )
+            )
+
+            site_type = cur.fetchone()[0]
+
+            # This is a site which is directly connected to the current site
+            # and will be implemented as a sub_tile. This information is relevant
+            # to generate correct placement constraints for sub_tiles directly
+            # connected within the same physical tile
+            connected_to_site = sites_with_direct.get(site_name, None)
```

Comment:
Rather than "None", should the default be ""?

Reply:
Actually `None` gets translated as an empty string in the csv, and read back as empty strings